### ISSUE TITLE
Proposal: PTHREAD_MUTEX_ERRORCHECK instead of PTHREAD_MUTEX_NORMAL

### ISSUE THREAD

**Issue by smklein** [rust-lang/rust#86575]
**TL;DR: Currently, double-locking a Mutex from Unix systems causes a deadlock. It doesn't have to! It totally could result in a (presumably easier-to-debug) panic with a fairly minor change.**

On Unix systems, the pthread mutex type has been chosen as "PTHREAD_MUTEX_NORMAL" within the implementation of Rust's [Mutex](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/unix/mutex.rs#L57) type:

```rust
        let mut attr = MaybeUninit::<libc::pthread_mutexattr_t>::uninit();
        cvt_nz(libc::pthread_mutexattr_init(attr.as_mut_ptr())).unwrap();
        let attr = PthreadMutexAttr(&mut attr);
        cvt_nz(libc::pthread_mutexattr_settype(attr.0.as_mut_ptr(), libc::PTHREAD_MUTEX_NORMAL))
            .unwrap();
        cvt_nz(libc::pthread_mutex_init(self.inner.get(), attr.0.as_ptr())).unwrap();
``` 

According to [Posix.1-2008's pthread_mutex_lock documentation](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_lock.html#), this has the behavior that:

> "Attempting to relock the mutex causes deadlock. If a thread attempts to unlock a mutex that it has not locked or a mutex which is unlocked, undefined behaviour results."

In contrast, if we opted for a different Mutex type - namely, ERRORCHECK:

```rust
        cvt_nz(libc::pthread_mutexattr_settype(attr.0.as_mut_ptr(), libc::PTHREAD_MUTEX_ERRORCHECK))
            .unwrap();
``` 

... This would result in the [behavior that](https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_mutex_lock.html):

> "If a thread attempts to relock a mutex that it has already locked, an error will be returned. If a thread attempts to unlock a mutex that it has not locked or a mutex which is unlocked, an error will be returned."

Within the context of the stdlib, the "returned error" actually means that we could fire an assertion when this behavior occurs, which IMO seems preferable to (on double locking) deadlocking or (on double unlock - an admittedly less plausible case) UB.

The [Mutex::lock documentation](https://doc.rust-lang.org/std/sync/struct.Mutex.html#panics) states:

> This function might panic when called if the lock is already held by the current thread.

Although I'm aware that "might" is load-bearing, making the change to `PTHREAD_MUTEX_ERRORCHECK` seems like it would be more-inline with the spirit of this documentation.

Some considerations:
- Is anyone actually dependent on the current deadlock behavior?
- Performance implications? Is there a benchmark suite used by Rust's stdlib we could use to gather some data here?

**Comment by joshtriplett** [comment]
Given the discussion summarized at https://github.com/rust-lang/rust/pull/86582#issuecomment-884402418 , closing this issue.

(I expect that future improvements to the standard library mutexes may address this.)
